README.md
# rio-glui

Explore CloudOptimized geotiff on your browser using Mapbox GL JS.

![](http://i.giphy.com/3ohzdVQrl8uUc8I2dO.gif)

## Installation

You can install rio-glui using pip

```
  $ pip install -U pip
  $ pip install rio-glui
```

or install from source:

```
$ git clone https://github.com/mapbox/rio-glui.git
$ cd rio-glui
$ pip install -e .
```

## Usage
```
Usage: rio glui [OPTIONS] PATH

  Rasterio glui cli.

Options:
  -b, --bidx BIDX                Raster band index (default: 1,2,3)
  --tiles-format [png|jpg|webp]  Tile image format (default: png)
  --tiles-dimensions INTEGER     Dimension of images being served (default:
                                 512)
  --nodata INTEGER               Force mask creation from a given nodata value
  --gl-tile-size INTEGER         mapbox-gl tileSize (default: 512)
  --port INTEGER                 Webserver port (default: 8080)
  --playground                   Launch playground app
  --mapbox-token TOKEN           Pass Mapbox token
  --help                         Show this message and exit.
```

Explore COG hosted on aws
```
 » rio glui https://oin-hotosm.s3.amazonaws.com/5ac626e091b5310010e0d482/0/5ac626e091b5310010e0d483.tif
Inspecting https://oin-hotosm.s3.amazonaws.com/5ac626e091b5310010e0d482/0/5ac626e091b5310010e0d483.tif at http://127.0.0.1:8080/index.html
```

### Playground

The `--playground` option opens a `playgroud` template where you an interact with the data to apply `rio-color formula`


#### Creating CloudOptimized Geotiff

To create rio-glui friendly files (CloudOptimized Geotiff) you can use another rasterio plugin: [`rio-cogeo`](https://github.com/mapbox/rio-cogeo.git)


#### Extras

This plugin also enable raster visualisation in Jupyter Notebook using [`mapboxgl-jupyter`](https://github.com/mapbox/mapboxgl-jupyter)


#### Contribution & Devellopement

Issues and pull requests are more than welcome.

###### Dev install & Pull-Request

```
$ git clone https://github.com/mapbox/rio-glui.git
$ cd rio-cogeo
$ pip install -e .[dev]
```

 *Python3.6 only*

This repo is set to use `pre-commit` to run *flake8*, *pydocstring* and *black* ("uncompromising Python code formatter") when commiting new code.

```
$ pre-commit install
$ git add .
$ git commit -m'my change'
black....................................................................Passed
Flake8...................................................................Passed
Verifying PEP257 Compliance..............................................Passed
$ git push origin
```


CHANGES.txt
1.0.6 (2019-02-14)
------------------
- update for rio-tiler >= 1.0 (#35)
- Add non-integer nodata option and add better test


tests/test_cli.py
"""tests rio_glui.server."""

import os
import pytest

import numpy
from mock import patch

from click.testing import CliRunner

from rio_glui.scripts.cli import glui

raster_path = os.path.join(
    os.path.dirname(__file__), "fixtures", "16-21560-29773_small_ycbcr.tif"
)

raster_ndvi_path = os.path.join(os.path.dirname(__file__), "fixtures", "ndvi_cogeo.tif")
raster_nodata_path = os.path.join(
    os.path.dirname(__file__), "fixtures", "internal_nodata.tif"
)


@pytest.fixture(autouse=True)
def testing_env_var(monkeypatch):
    """Set testing env variable."""
    monkeypatch.delenv("MAPBOX_ACCESS_TOKEN", raising=False)


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_valid(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(glui, [raster_path])
    TileServer.assert_called_once()
    assert not result.exception
    assert result.exit_code == 0


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validbidx(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(glui, [raster_path, "--bidx", "1,2,3"])
    TileServer.assert_called_once()
    assert not result.exception
    assert result.exit_code == 0


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_valid_playground(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_playround_url.return_value = (
        "http://127.0.0.1:8080/playground.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(glui, [raster_path, "--playground"])
    TileServer.assert_called_once()
    assert not result.exception
    assert result.exit_code == 0


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validEnvToken(launch, TileServer, monkeypatch):
    """Should work as expected."""
    monkeypatch.setenv("MAPBOX_ACCESS_TOKEN", "pk.afakemapboxtoken")

    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(glui, [raster_path])
    TileServer.assert_called_once()
    assert not result.exception
    assert result.exit_code == 0


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validToken(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(glui, [raster_path, "--mapbox-token", "pk.afakemapboxtoken"])
    TileServer.assert_called_once()
    assert not result.exception
    assert result.exit_code == 0


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validInvalidToken(launch, TileServer):
    """Should work as expected."""
    runner = CliRunner()
    result = runner.invoke(glui, [raster_path, "--mapbox-token", "sk.afakemapboxtoken"])
    TileServer.assert_not_called()
    launch.assert_not_called()
    assert result.exception
    assert result.exit_code == 1


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validInvalidbidx(launch, TileServer):
    """Should work as expected."""
    runner = CliRunner()
    result = runner.invoke(glui, [raster_path, "--bidx", "1,a,2"])
    TileServer.assert_not_called()
    launch.assert_not_called()
    assert result.exception
    assert result.exit_code == 1


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validInvalidbidxLength(launch, TileServer):
    """Should work as expected."""
    runner = CliRunner()
    result = runner.invoke(glui, [raster_path, "--bidx", "1,2"])
    TileServer.assert_not_called()
    launch.assert_not_called()
    assert result.exception
    assert result.exit_code == 1


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validInvalidbidxZero(launch, TileServer):
    """Should work as expected."""
    runner = CliRunner()
    result = runner.invoke(glui, [raster_path, "--bidx", "0"])
    TileServer.assert_not_called()
    launch.assert_not_called()
    assert result.exception
    assert result.exit_code == 1


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validScaleColorMap(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(
        glui, [raster_ndvi_path, "--scale", "-1", "1", "--colormap", "cfastie"]
    )
    TileServer.assert_called_once()
    assert not result.exception
    assert result.exit_code == 0


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_invalidScale(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(glui, [raster_ndvi_path, "--scale", "-1"])
    TileServer.assert_not_called()
    launch.assert_not_called()
    assert result.exception
    assert result.exit_code == 2


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_invalidScaleNumber(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(
        glui, [raster_ndvi_path, "--scale", "-1", "1", "--scale", "0", "1"]
    )
    TileServer.assert_not_called()
    launch.assert_not_called()
    assert result.exception
    assert result.exit_code == 1


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validNodata(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(
        glui, [raster_nodata_path, "--bidx", "1", "--nodata", "-9999"]
    )
    TileServer.assert_called_once()
    raster = TileServer.call_args[0][0]
    assert raster.nodata == -9999
    assert not result.exception
    assert result.exit_code == 0


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validNodataNone(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(glui, [raster_path, "--bidx", "1", "--nodata", "None"])
    TileServer.assert_called_once()
    raster = TileServer.call_args[0][0]
    assert not raster.nodata
    assert not result.exception
    assert result.exit_code == 0


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validNodataNan(launch, TileServer):
    """Should work as expected."""
    TileServer.return_value.get_template_url.return_value = (
        "http://127.0.0.1:8080/index.html"
    )
    TileServer.return_value.start.return_value = True

    launch.return_value = True

    runner = CliRunner()
    result = runner.invoke(glui, [raster_nodata_path, "--bidx", "1", "--nodata", "nan"])
    TileServer.assert_called_once()
    raster = TileServer.call_args[0][0]
    assert not numpy.isfinite(raster.nodata)
    assert not result.exception
    assert result.exit_code == 0


@patch("rio_glui.server.TileServer")
@patch("click.launch")
def test_glui_validBadNodata(launch, TileServer):
    """Should work as expected."""
    runner = CliRunner()
    result = runner.invoke(
        glui, [raster_nodata_path, "--bidx", "1", "--nodata", "jacques"]
    )
    TileServer.assert_not_called()
    launch.assert_not_called()
    assert result.exception
    assert result.exit_code == 1


tests/test_raster.py
"""tests rio_glui.raster."""

import os
import pytest

from rio_glui.raster import RasterTiles, _meters_per_pixel

raster_path = os.path.join(
    os.path.dirname(__file__), "fixtures", "16-21560-29773_small_ycbcr.tif"
)
invalid_raster_path = os.path.join(
    os.path.dirname(__file__), "fixtures", "16-21560-29773_small.tif"
)
raster_nodata_path = os.path.join(
    os.path.dirname(__file__), "fixtures", "internal_nodata.tif"
)


def test_meters_per_pixel_valid():
    """Should work as expected (returns correct pixel size)."""
    assert _meters_per_pixel(18, 45) == 0.4222589143296568
    assert _meters_per_pixel(18, -45) == 0.4222589143296568


def test_rastertiles_valid():
    """Should work as expected (create rastertiles object)."""
    r = RasterTiles(raster_path)
    assert r.path == raster_path
    assert r.tiles_size == 512
    assert not r.nodata
    assert r.bounds == [
        -61.56738281249997,
        16.225223624120076,
        -61.5618896507246,
        16.23049792684362,
    ]
    assert r.indexes == (1, 2, 3)
    assert r.overiew_levels == [2, 4, 8, 16, 32, 64]


def test_rastertiles_invalidcogeo():
    """Should error with invalid Cogeo format."""
    with pytest.raises(Exception):
        RasterTiles(invalid_raster_path)


def test_rastertiles_valid_indexes_option():
    """Should work as expected (create rastertiles object)."""
    r = RasterTiles(raster_path, indexes=[1])
    assert r.path == raster_path
    assert r.indexes == [1]
    assert r.tiles_size == 512
    assert not r.nodata


def test_rastertiles_valid_size_option():
    """Should work as expected (create rastertiles object)."""
    r = RasterTiles(raster_path, tiles_size=256)
    assert r.path == raster_path
    assert r.tiles_size == 256
    assert not r.nodata


def test_rastertiles_valid_nodata_option():
    """Should work as expected (create rastertiles object)."""
    r = RasterTiles(raster_path, nodata=0)
    assert r.path == raster_path
    assert r.nodata == 0


def test_rastertiles_valid_nodata():
    """Should work as expected (create rastertiles object)."""
    r = RasterTiles(raster_nodata_path)
    assert r.path == raster_nodata_path
    assert r.nodata == -9999


def test_rastertiles_get_bounds():
    """Should work as expected (create rastertiles object and get bounds)."""
    r = RasterTiles(raster_path)
    assert r.get_bounds() == [
        -61.56738281249997,
        16.225223624120076,
        -61.5618896507246,
        16.23049792684362,
    ]


def test_rastertiles_get_centers():
    """Should work as expected (create rastertiles object and get center)."""
    r = RasterTiles(raster_path)
    assert r.get_center() == [-61.56463623161228, 16.227860775481847]


def test_rastertiles_tile_exists_valid():
    """Should work as expected (create rastertiles object and check if tile exists)."""
    r = RasterTiles(raster_path)
    z = 18
    x = 86240
    y = 119094
    assert r.tile_exists(z, x, y)


def test_rastertiles_tile_exists_false():
    """Should work as expected (create rastertiles object and check if tile exists)."""
    r = RasterTiles(raster_path)
    z = 18
    x = 8240
    y = 119094
    assert not r.tile_exists(z, x, y)


def test_rastertiles_get_max_zoom():
    """Should work as expected (create rastertiles object and get_max_zoom)."""
    r = RasterTiles(raster_path)
    assert r.get_max_zoom() == 19


def test_rastertiles_get_min_zoom():
    """Should work as expected (create rastertiles object and get_min_zoom)."""
    r = RasterTiles(raster_path)
    assert r.get_min_zoom() == 12


def test_rastertiles_read_tile():
    """Should work as expected (create rastertiles object and read tile)."""
    r = RasterTiles(raster_path)
    z = 18
    x = 86240
    y = 119094
    data, mask = r.read_tile(z, x, y)
    assert data.shape == (3, 512, 512)
    assert mask.shape == (512, 512)


def test_rastertiles_read_tile_small():
    """Should work as expected (create rastertiles object and read tile)."""
    r = RasterTiles(raster_path, tiles_size=256)
    z = 18
    x = 86240
    y = 119094
    data, mask = r.read_tile(z, x, y)
    assert data.shape == (3, 256, 256)
    assert mask.shape == (256, 256)


tests/test_server.py
"""tests rio_glui.server."""

import os

import numpy
from tornado.testing import AsyncHTTPTestCase

import mercantile
from rio_tiler.utils import tile_read

from rio_glui.raster import RasterTiles
from rio_glui.server import TileServer

raster_path = os.path.join(
    os.path.dirname(__file__), "fixtures", "16-21560-29773_small_ycbcr.tif"
)
raster_ndvi_path = os.path.join(os.path.dirname(__file__), "fixtures", "ndvi_cogeo.tif")
invalid_raster_path = os.path.join(
    os.path.dirname(__file__), "fixtures", "16-21560-29773_small.tif"
)


def test_TileServer_default():
    """Should work as expected (create TileServer object)."""
    r = RasterTiles(raster_path)
    app = TileServer(r)
    assert app.raster == r
    assert app.port == 8080
    assert not app.server
    assert app.tiles_format == "png"
    assert app.gl_tiles_size == 512
    assert app.gl_tiles_minzoom == 0
    assert app.gl_tiles_maxzoom == 22


def test_TileServer_1b():
    """Should work as expected (create TileServer object)."""
    r = RasterTiles(raster_ndvi_path)
    app = TileServer(
        r,
        tiles_format="jpg",
        gl_tiles_minzoom=13,
        gl_tiles_maxzoom=19,
        gl_tiles_size=256,
        scale=((-1, 1),),
        colormap="cfastie",
        port=5000,
    )
    assert app.raster == r
    assert app.port == 5000
    assert not app.server
    assert app.tiles_format == "jpg"
    assert app.gl_tiles_size == 256
    assert app.gl_tiles_minzoom == 13
    assert app.gl_tiles_maxzoom == 19


def test_TileServer_options():
    """Should work as expected (create TileServer object)."""
    r = RasterTiles(raster_path)
    app = TileServer(
        r,
        tiles_format="jpg",
        gl_tiles_minzoom=13,
        gl_tiles_maxzoom=19,
        gl_tiles_size=256,
        port=5000,
    )
    assert app.raster == r
    assert app.port == 5000
    assert not app.server
    assert app.tiles_format == "jpg"
    assert app.gl_tiles_size == 256
    assert app.gl_tiles_minzoom == 13
    assert app.gl_tiles_maxzoom == 19


def test_TileServer_raster_tilesize():
    """Should work as expected (create TileServer object)."""
    r = RasterTiles(raster_path, tiles_size=256)
    app = TileServer(r)
    assert app.raster == r
    assert not app.server
    assert app.tiles_format == "png"
    assert app.gl_tiles_size == 256


def test_TileServer_raster_get_bounds():
    """Should work as expected."""
    r = RasterTiles(raster_path)
    app = TileServer(r)
    assert app.raster == r
    assert app.get_bounds() == r.get_bounds()


def test_TileServer_raster_get_center():
    """Should work as expected."""
    r = RasterTiles(raster_path)
    app = TileServer(r)
    assert app.raster == r
    assert app.get_center() == r.get_center()


def test_TileServer_get_tiles_url():
    """Should work as expected (create TileServer object and get tiles endpoint)."""
    r = RasterTiles(raster_path)
    app = TileServer(r)
    assert app.get_tiles_url() == "http://127.0.0.1:8080/tiles/{z}/{x}/{y}.png"


def test_TileServer_get_template_url():
    """Should work as expected (create TileServer object and get template url)."""
    r = RasterTiles(raster_path)
    app = TileServer(r)
    assert app.get_template_url() == "http://127.0.0.1:8080/index.html"


def test_TileServer_get_playground_url():
    """Should work as expected (create TileServer object and get playground url)."""
    r = RasterTiles(raster_path)
    app = TileServer(r)
    assert app.get_playground_url() == "http://127.0.0.1:8080/playground.html"


class TestHandlers(AsyncHTTPTestCase):
    """Test tornado handlers."""

    def get_app(self):
        """Initialize app."""
        r = RasterTiles(raster_path)
        return TileServer(r).app

    def test_get_root(self):
        """Should return error on root query."""
        response = self.fetch("/")
        self.assertEqual(response.code, 404)

    def test_tile(self):
        """Should return tile buffer."""
        response = self.fetch("/tiles/18/86240/119094.png")
        self.assertEqual(response.code, 200)
        self.assertTrue(response.buffer)
        self.assertEqual(response.headers["Content-Type"], "image/png")

    def test_tileColor(self):
        """Should apply color ops and return tile buffer."""
        response = self.fetch("/tiles/18/86240/119094.png?color=gamma%20b%201.8")
        self.assertEqual(response.code, 200)
        self.assertTrue(response.buffer)
        self.assertEqual(response.headers["Content-Type"], "image/png")

    def test_tileJpeg(self):
        """Should return tile jpeg buffer."""
        response = self.fetch("/tiles/18/86240/119094.jpg")
        self.assertEqual(response.code, 200)
        self.assertTrue(response.buffer)
        self.assertEqual(response.headers["Content-Type"], "image/jpg")

    def test_tileNotFound(self):
        """Should error with tile doesn't exits."""
        response = self.fetch("/tiles/18/8624/119094.png")
        self.assertEqual(response.code, 404)

    def test_TemplateSimple(self):
        """Should find the template."""
        response = self.fetch("/index.html")
        self.assertEqual(response.code, 200)

    def test_TemplatePlayground(self):
        """Should find the template."""
        response = self.fetch("/playground.html")
        self.assertEqual(response.code, 200)


class TestHandlersRescale(AsyncHTTPTestCase):
    """Test tornado handlers."""

    def get_app(self):
        """Initialize app."""
        r = RasterTiles(raster_path)
        return TileServer(r, scale=((1, 240),)).app

    def test_tile(self):
        """Should return tile buffer."""
        response = self.fetch("/tiles/18/86240/119094.png")
        self.assertEqual(response.code, 200)
        self.assertTrue(response.buffer)
        self.assertEqual(response.headers["Content-Type"], "image/png")


class TestHandlers1B(AsyncHTTPTestCase):
    """Test tornado handlers."""

    def get_app(self):
        """Initialize app."""
        r = RasterTiles(raster_ndvi_path, tiles_size=32)
        return TileServer(r, gl_tiles_size=32, scale=((-1, 1),), colormap="cfastie").app

    def test_tile(self):
        """Should return tile buffer."""
        response = self.fetch("/tiles/9/142/205.png")
        self.assertEqual(response.code, 200)
        self.assertTrue(response.buffer)
        self.assertEqual(response.headers["Content-Type"], "image/png")


class CustomRaster(RasterTiles):
    """Custom RasterTiles."""

    def read_tile(self, z, x, y):
        """Read raster tile data and mask."""
        mercator_tile = mercantile.Tile(x=x, y=y, z=z)
        tile_bounds = mercantile.xy_bounds(mercator_tile)

        data, mask = tile_read(
            self.path,
            tile_bounds,
            self.tiles_size,
            indexes=self.indexes,
            nodata=self.nodata,
        )
        data = (data[0] + data[1]) / 2
        return data.astype(numpy.uint8), mask


class TestHandlersCustom(AsyncHTTPTestCase):
    """Test tornado handlers."""

    def get_app(self):
        """Initialize app."""
        r = CustomRaster(raster_path)
        return TileServer(r).app

    def test_tile(self):
        """Should return tile buffer."""
        response = self.fetch("/tiles/18/86240/119094.png")
        self.assertEqual(response.code, 200)
        self.assertTrue(response.buffer)
        self.assertEqual(response.headers["Content-Type"], "image/png")


rio_glui/raster.py
"""rio_glui.raster: raster tiles object."""

import math

import mercantile
import rasterio
from rasterio.warp import transform_bounds, calculate_default_transform

from rio_tiler.utils import tile_read


def _meters_per_pixel(zoom, lat):
    return (math.cos(lat * math.pi / 180.0) * 2 * math.pi * 6378137) / (256 * 2 ** zoom)


class RasterTiles(object):
    """
    Raster tiles object.

    Attributes
    ----------
    src_path : str or PathLike object
        A dataset path or URL. Will be opened in "r" mode.
    indexes : tuple, int, optional
        Raster band indexes to read.
    tiles_size: int, optional (default: 512)
        X/Y tile size to return.
    nodata: int, optional
        nodata value for mask creation.

    Methods
    -------
    get_bounds()
        Get raster bounds (WGS84).
    get_center()
        Get raster lon/lat center coordinates.
    tile_exists(z, x, y)
        Check if a mercator tile is within raster bounds.
    get_max_zoom(snap=0.5, max_z=23)
        Calculate raster max zoom level.
    get_min_zoom(snap=0.5, max_z=23)
        Calculate raster min zoom level.
    read_tile( z, x, y)
        Read raster tile data and mask.

    """

    def __init__(self, src_path, indexes=None, tiles_size=512, nodata=None):
        """Initialize RasterTiles object."""
        self.path = src_path
        self.tiles_size = tiles_size
        with rasterio.open(src_path) as src:
            try:
                assert src.driver == "GTiff"
                assert src.is_tiled
                assert src.overviews(1)
            except (AttributeError, AssertionError, KeyError):
                raise Exception(
                    "{} is not a valid CloudOptimized Geotiff".format(src_path)
                )

            self.bounds = list(
                transform_bounds(
                    *[src.crs, "epsg:4326"] + list(src.bounds), densify_pts=0
                )
            )
            self.indexes = indexes if indexes is not None else src.indexes
            self.nodata = nodata if nodata is not None else src.nodata
            self.crs = src.crs
            self.crs_bounds = src.bounds
            self.meta = src.meta
            self.overiew_levels = src.overviews(1)

    def get_bounds(self):
        """Get raster bounds (WGS84)."""
        return self.bounds

    def get_center(self):
        """Get raster lon/lat center coordinates."""
        lat = (self.bounds[3] - self.bounds[1]) / 2 + self.bounds[1]
        lng = (self.bounds[2] - self.bounds[0]) / 2 + self.bounds[0]
        return [lng, lat]

    def tile_exists(self, z, x, y):
        """Check if a mercator tile is within raster bounds."""
        mintile = mercantile.tile(self.bounds[0], self.bounds[3], z)
        maxtile = mercantile.tile(self.bounds[2], self.bounds[1], z)
        return (
            (x <= maxtile.x + 1)
            and (x >= mintile.x)
            and (y <= maxtile.y + 1)
            and (y >= mintile.y)
        )

    def get_max_zoom(self, snap=0.5, max_z=23):
        """Calculate raster max zoom level."""
        dst_affine, w, h = calculate_default_transform(
            self.crs,
            "epsg:3857",
            self.meta["width"],
            self.meta["height"],
            *self.crs_bounds
        )

        res_max = max(abs(dst_affine[0]), abs(dst_affine[4]))

        tgt_z = max_z
        mpp = 0.0

        # loop through the pyramid to file the closest z level
        for z in range(1, max_z):
            mpp = _meters_per_pixel(z, 0)

            if (mpp - ((mpp / 2) * snap)) < res_max:
                tgt_z = z
                break

        return tgt_z

    def get_min_zoom(self, snap=0.5, max_z=23):
        """Calculate raster min zoom level."""
        dst_affine, w, h = calculate_default_transform(
            self.crs,
            "epsg:3857",
            self.meta["width"],
            self.meta["height"],
            *self.crs_bounds
        )

        res_max = max(abs(dst_affine[0]), abs(dst_affine[4]))
        max_decim = self.overiew_levels[-1]
        resolution = max_decim * res_max

        tgt_z = 0
        mpp = 0.0

        # loop through the pyramid to file the closest z level
        for z in list(range(0, 24))[::-1]:
            mpp = _meters_per_pixel(z, 0)
            tgt_z = z

            if (mpp - ((mpp / 2) * snap)) > resolution:
                break

        return tgt_z

    def read_tile(self, z, x, y):
        """Read raster tile data and mask."""
        mercator_tile = mercantile.Tile(x=x, y=y, z=z)
        tile_bounds = mercantile.xy_bounds(mercator_tile)
        return tile_read(
            self.path,
            tile_bounds,
            self.tiles_size,
            indexes=self.indexes,
            nodata=self.nodata,
        )


rio_glui/scripts/cli.py
"""rio_glui.cli."""

import os
import click
import numpy

from rio_glui.raster import RasterTiles
from rio_glui import server


class MbxTokenType(click.ParamType):
    """Mapbox token type."""

    name = "token"

    def convert(self, value, param, ctx):
        """Validate token."""
        try:
            if not value:
                return ""

            assert value.startswith("pk")
            return value

        except (AttributeError, AssertionError):
            raise click.ClickException(
                "Mapbox access token must be public (pk). "
                "Please sign up at https://www.mapbox.com/signup/ to get a public token. "
                "If you already have an account, you can retreive your "
                "token at https://www.mapbox.com/account/."
            )


class BdxParamType(click.ParamType):
    """Band inddex type."""

    name = "bidx"

    def convert(self, value, param, ctx):
        """Validate and parse band index."""
        try:
            bands = [int(x) for x in value.split(",")]
            assert len(bands) in [1, 3]
            assert all(b > 0 for b in bands)
            return bands

        except (ValueError, AttributeError, AssertionError):
            raise click.ClickException(
                "bidx must be a string with 1 or 3 ints comma-separated, "
                "representing the band indexes for R,G,B"
            )


class NodataParamType(click.ParamType):
    """Nodata inddex type."""

    name = "nodata"

    def convert(self, value, param, ctx):
        """Validate and parse band index."""
        try:
            if value.lower() == "nan":
                return numpy.nan
            elif value.lower() in ["nil", "none", "nada"]:
                return None
            else:
                return float(value)
        except (TypeError, ValueError):
            raise click.ClickException("{} is not a valid nodata value.".format(value))


@click.command()
@click.argument("path", type=str)
@click.option("--bidx", "-b", type=BdxParamType(), help="Raster band index")
@click.option(
    "--scale",
    type=int,
    multiple=True,
    nargs=2,
    help="Min and Max data bounds to rescale data from. "
    "Form multiband you can either provide use '--scale 0 1000' or "
    "'--scale 0 1000 --scale 0 500 --scale 0 1500'",
)
@click.option(
    "--colormap",
    type=click.Choice(["cfastie", "schwarzwald"]),
    help=" Rio-tiler compatible colormap name",
)
@click.option(
    "--tiles-format",
    type=click.Choice(["png", "jpg", "webp"]),
    default="png",
    help="Tile image format (default: png)",
)
@click.option(
    "--tiles-dimensions",
    type=int,
    default=512,
    help="Dimension of images being served (default: 512)",
)
@click.option(
    "--nodata",
    type=NodataParamType(),
    metavar="NUMBER|nan",
    help="Set nodata masking values for input dataset.",
)
@click.option(
    "--gl-tile-size",
    type=int,
    help="mapbox-gl tileSize (default is the same as `tiles-dimensions`)",
)
@click.option("--port", type=int, default=8080, help="Webserver port (default: 8080)")
@click.option("--playground", is_flag=True, help="Launch playground app")
@click.option(
    "--mapbox-token",
    type=MbxTokenType(),
    metavar="TOKEN",
    default=lambda: os.environ.get("MAPBOX_ACCESS_TOKEN", ""),
    help="Pass Mapbox token",
)
def glui(
    path,
    bidx,
    scale,
    colormap,
    tiles_format,
    tiles_dimensions,
    nodata,
    gl_tile_size,
    port,
    playground,
    mapbox_token,
):
    """Rasterio glui cli."""
    if scale and len(scale) not in [1, 3]:
        raise click.ClickException("Invalid number of scale values")

    raster = RasterTiles(path, indexes=bidx, tiles_size=tiles_dimensions, nodata=nodata)

    app = server.TileServer(
        raster,
        scale=scale,
        colormap=colormap,
        tiles_format=tiles_format,
        gl_tiles_size=gl_tile_size,
        gl_tiles_minzoom=raster.get_min_zoom(),
        gl_tiles_maxzoom=raster.get_max_zoom(),
        port=port,
    )

    if playground:
        url = app.get_playground_url()
    else:
        url = app.get_template_url()

    if mapbox_token:
        url = "{}?access_token={}".format(url, mapbox_token)

    click.launch(url)
    click.echo("Inspecting {} at {}".format(path, url), err=True)
    app.start()
